### PR TITLE
Fix: Implement missing duckdb to vortex type conversions

### LINKED ISSUE #4848: crash when using list fields
```
$ RUST_BACKTRACE=1 duckdb 
DuckDB v1.4.0 (Andium) b8a06e4a22
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D load vortex;
D copy (select ['a']) to 'test.vortex' (FORMAT VORTEX);

thread '<unnamed>' panicked at vortex-duckdb/src/convert/dtype.rs:245:46:
not yet implemented
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/29483883eed69d5fb4db01964cdf2af4d86e9cb2/library/std/src/panicking.rs:697:5
   1: core::panicking::panic_fmt
             at /rustc/29483883eed69d5fb4db01964cdf2af4d86e9cb2/library/core/src/panicking.rs:75:14
   2: core::panicking::panic
             at /rustc/29483883eed69d5fb4db01964cdf2af4d86e9cb2/library/core/src/panicking.rs:145:5
   3: <vortex_dtype::dtype::DType as vortex_duckdb::convert::dtype::FromLogicalType>::from_logical_type
   4: vortex_duckdb::duckdb::copy_function::callback::bind_callback
   5: _ZN6vortex10c_bind_oneERN6duckdb13ClientContextERNS0_21CopyFunctionBindInputERKNS0_6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEELb1EEERKNS5_INS0_11LogicalTypeELb1EEE
   6: _ZN6duckdb6Binder10BindCopyToERNS_13CopyStatementERKNS_12CopyFunctionENS_10CopyToTypeE
   7: _ZN6duckdb6Binder4BindERNS_13CopyStatementENS_10CopyToTypeE
   8: _ZN6duckdb6Binder4BindERNS_12SQLStatementE
   9: _ZN6duckdb7Planner10CreatePlanERNS_12SQLStatementE
  10: _ZN6duckdb13ClientContext31CreatePreparedStatementInternalERNS_17ClientContextLockERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_10unique_ptrINS_12SQLStatementESt14default_deleteISC_ELb1EEENS_12optional_ptrISt13unordered_mapIS8_NS_18BoundParameterDataENS_33CaseInsensitiveStringHashFunctionENS_29CaseInsensitiveStringEqualityESaISt4pairIS9_SI_EEELb1EEE
  11: _ZN6duckdb13ClientContext23CreatePreparedStatementERNS_17ClientContextLockERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_10unique_ptrINS_12SQLStatementESt14default_deleteISC_ELb1EEENS_12optional_ptrISt13unordered_mapIS8_NS_18BoundParameterDataENS_33CaseInsensitiveStringHashFunctionENS_29CaseInsensitiveStringEqualityESaISt4pairIS9_SI_EEELb1EEENS_21PreparedStatementModeE
  12: _ZN6duckdb13ClientContext24PendingStatementInternalERNS_17ClientContextLockERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_10unique_ptrINS_12SQLStatementESt14default_deleteISC_ELb1EEERKNS_22PendingQueryParametersE
  13: _ZN6duckdb13ClientContext35PendingStatementOrPreparedStatementERNS_17ClientContextLockERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_10unique_ptrINS_12SQLStatementESt14default_deleteISC_ELb1EEERNS_10shared_ptrINS_21PreparedStatementDataELb1EEERKNS_22PendingQueryParametersE
  14: _ZN6duckdb13ClientContext43PendingStatementOrPreparedStatementInternalERNS_17ClientContextLockERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_10unique_ptrINS_12SQLStatementESt14default_deleteISC_ELb1EEERNS_10shared_ptrINS_21PreparedStatementDataELb1EEERKNS_22PendingQueryParametersE
  15: _ZN6duckdb13ClientContext20PendingQueryInternalERNS_17ClientContextLockENS_10unique_ptrINS_12SQLStatementESt14default_deleteIS4_ELb1EEERKNS_22PendingQueryParametersEb
  16: _ZN6duckdb13ClientContext12PendingQueryENS_10unique_ptrINS_12SQLStatementESt14default_deleteIS2_ELb1EEERSt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_18BoundParameterDataENS_33CaseInsensitiveStringHashFunctionENS_29CaseInsensitiveStringEqualityESaISt4pairIKSC_SD_EEEb
  17: _ZN6duckdb13ClientContext12PendingQueryENS_10unique_ptrINS_12SQLStatementESt14default_deleteIS2_ELb1EEEb
  18: _ZN6duckdb10Connection12PendingQueryENS_10unique_ptrINS_12SQLStatementESt14default_deleteIS2_ELb1EEEb
  19: duckdb_shell_sqlite3_prepare_v2
  20: _ZN12duckdb_shell10ShellState10ExecuteSQLEPKcPPc
  21: _ZN12duckdb_shell10ShellState13RunOneSqlLineENS_9InputModeEPc
  22: _ZN12duckdb_shell10ShellState12ProcessInputENS_9InputModeE
  23: main
  24: __libc_start_call_main
             at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
  25: __libc_start_main_impl
             at ./csu/../csu/libc-start.c:392:3
  26: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
fatal runtime error: failed to initiate panic, error 5, aborting
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by robert3005**
Implement StructExporter, List and Struct scalar conversions and Struct and list type conversions from duckdb to vortex, fix https://github.com/vortex-data/vortex/issues/4848

Signed-off-by: Robert Kruszewski <github@robertk.io>

### CODE REVIEW

File: vortex-duckdb/src/duckdb/logical_type.rs [hunk 1]
```diff
@@ -199,8 +187,9 @@ impl LogicalType {
         unsafe { LogicalType::own(duckdb_array_type_child_type(self.as_ptr())) }
     }
 
-    pub fn array_type_array_size(&self) -> idx_t {
-        unsafe { duckdb_array_type_array_size(self.as_ptr()) }
+    pub fn array_type_array_size(&self) -> u32 {
+        u32::try_from(unsafe { duckdb_array_type_array_size(self.as_ptr()) })
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/duckdb/logical_type.rs:200]: why?

File: vortex-duckdb/src/duckdb/logical_type.rs [hunk 2]
```diff
@@ -215,26 +204,85 @@ impl LogicalType {
...
-        })?;
+        write!(f, "{}", unsafe { CStr::from_ptr(debug).to_string_lossy() })?;
         unsafe { duckdb_free(debug.cast()) };
         Ok(())
     }
 }
 
+/// String wrapper that represents strings returned from duckdb C API. These need to be freed using `duckdb_free`.
+#[derive(Debug)]
+pub struct DDBString(*mut std::ffi::c_char);
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/duckdb/logical_type.rs]: why can we not sure the wrapper macro here?

File: vortex-duckdb/src/duckdb/logical_type.rs [hunk 3]
```diff
@@ -215,26 +204,85 @@ impl LogicalType {
...
+    pub fn new(ptr: *mut std::ffi::c_char) -> Self {
+        Self(ptr)
+    }
+}
+
+impl AsRef<str> for DDBString {
+    fn as_ref(&self) -> &str {
+        unsafe { CStr::from_ptr(self.0) }
+            .to_str()
+            .map_err(|e| vortex_err!("Failed to convert CStr to str: {e}"))
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/duckdb/logical_type.rs]: there is an infallible lossy convention. do we want that for the from?

File: vortex-duckdb/src/exporter/fixed_size_list.rs [hunk 4]
```diff
@@ -102,7 +103,11 @@ mod tests {
         offset: usize,
         len: usize,
     ) -> DataChunk {
-        let array_type = LogicalType::new_array(cpp::DUCKDB_TYPE::DUCKDB_TYPE_INTEGER, list_size);
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/exporter/fixed_size_list.rs:105]: would be nice to keep this array ctor. as it's the duckdb terminology

File: vortex-duckdb/src/convert/dtype.rs [hunk 5]
```diff
@@ -116,21 +121,45 @@ impl FromLogicalType for DType {
...
+                nullability,
+            ),
+            DUCKDB_TYPE::DUCKDB_TYPE_LIST => DType::List(
+                Arc::new(DType::from_logical_type(
+                    logical_type.list_child_type(),
+                    Nullability::Nullable,
+                )?),
+                nullability,
+            ),
+            DUCKDB_TYPE::DUCKDB_TYPE_STRUCT => DType::Struct(
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/convert/dtype.rs:139]: ooi. why impl this without the exporter?

File: vortex-duckdb/src/duckdb/logical_type.rs (hunk 1 above)
Comment (robert3005) [reply vortex-duckdb/src/duckdb/logical_type.rs:200]: our fixedsizelist has this limitation so all usages in our codebase will have to do it

File: vortex-duckdb/src/convert/dtype.rs (hunk 5 above)
Comment (robert3005) [reply vortex-duckdb/src/convert/dtype.rs:139]: I wanted it for the importer but I see things will go badly if one is implemented without the other

File: vortex-duckdb/src/exporter/fixed_size_list.rs (hunk 4 above)
Comment (robert3005) [reply vortex-duckdb/src/exporter/fixed_size_list.rs:105]: we can keep this, I thought it was redundant

File: vortex-duckdb/src/duckdb/logical_type.rs (hunk 2 above)
Comment (robert3005) [reply vortex-duckdb/src/duckdb/logical_type.rs]: we can, didn't think the macro was that helpful here

Comment (robert3005) [reply vortex-duckdb/src/duckdb/logical_type.rs]: I don't think the macro is that useful here, I would have to change it to support this case and it's a singular use which would complicate it needlessly

File: vortex-duckdb/src/exporter/struct_.rs
Comment (joseph-isaacs) [inline vortex-duckdb/src/exporter/struct_.rs:1]: can we use `struct_`?

File: vortex-duckdb/src/exporter/struct_.rs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        .children()
+        .into_iter()
+        .map(|child| new_array_exporter(&child, cache))
+        .collect::<VortexResult<Vec<_>>>()?;
+    let struct_exporter = Box::new(StructExporter { children });
+    Ok(if array.dtype().is_nullable() {
+        validity::new_exporter(array.validity_mask(), struct_exporter)
+    } else {
+        struct_exporter
+    })
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/exporter/struct_.rs:40]: might be worth explaining how duckdb handles nested nulls here?

File: vortex-duckdb/src/exporter/struct_.rs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    use vortex::IntoArray;
+    use vortex::arrays::{PrimitiveArray, VarBinViewArray};
+    use vortex::error::{VortexExpect, VortexUnwrap};
+
+    use super::*;
+    use crate::cpp;
+    use crate::duckdb::{DataChunk, LogicalType};
+
+    #[test]
+    fn test_struct_exporter() {
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/exporter/struct_.rs:69]: can we add a test with null in the both parent and child?

File: vortex-duckdb/src/duckdb/logical_type.rs [hunk 8]
```diff
@@ -215,26 +213,91 @@ impl LogicalType {
...
+            .to_str()
+            .map_err(|e| vortex_err!("Failed to convert C string to str {e}"))?;
+
+        Ok(Self(ptr))
+    }
+}
+
+impl AsRef<str> for DDBString {
+    fn as_ref(&self) -> &str {
+        // SAFETY: The string have been validated on construction.
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/duckdb/logical_type.rs:271]: I dont think this is necessarily validated in duckdb

File: vortex-duckdb/src/duckdb/logical_type.rs [hunk 9]
```diff
@@ -215,26 +213,91 @@ impl LogicalType {
...
-        })?;
+        write!(f, "{}", unsafe { CStr::from_ptr(debug).to_string_lossy() })?;
         unsafe { duckdb_free(debug.cast()) };
         Ok(())
     }
 }
 
+/// String wrapper that represents strings returned from duckdb C API. These need to be freed using `duckdb_free`.
+#[derive(Debug)]
+pub struct DDBString(*mut std::ffi::c_char);
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/duckdb/logical_type.rs]: Why not this
```suggestion
wrapper!(
    DDBString,
    *mut std::ffi::c_char,
    |ptr: &mut *mut std::ffi::c_char| unsafe { duckdb_free((*ptr).cast()) }
);```

Comment (joseph-isaacs) [reply vortex-duckdb/src/duckdb/logical_type.rs]: I would really like to use the wrapper if possible?

File: vortex-duckdb/src/exporter/fixed_size_list.rs [hunk 10]
```diff
@@ -282,13 +283,9 @@ mod tests {
...
             LogicalType::new_array(cpp::DUCKDB_TYPE::DUCKDB_TYPE_INTEGER, inner_list_size);
-        // SAFETY: inner_array_type is a valid LogicalType created above.
-        unsafe {
-            LogicalType::own(cpp::duckdb_create_array_type(
-                inner_array_type.as_ptr(),
-                outer_list_size as cpp::idx_t,
-            ))
-        }
+
+        LogicalType::fixed_size_list_type(inner_array_type, outer_list_size)
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/exporter/fixed_size_list.rs]: Would be nice if this was called `LogicalType::array_type`

File: vortex-duckdb/src/duckdb/logical_type.rs
Comment (joseph-isaacs) [inline vortex-duckdb/src/duckdb/logical_type.rs:1]: Would be nice if we can keep with most vortex agnostic

File: vortex-duckdb/src/duckdb/logical_type.rs (hunk 8 above)
Comment (robert3005) [reply vortex-duckdb/src/duckdb/logical_type.rs:271]: but it's validated when we construct new instance of this wrapper, if we use a macro we would have to validate on every invocation

File: vortex-duckdb/src/duckdb/logical_type.rs (hunk 9 above)
Comment (robert3005) [reply vortex-duckdb/src/duckdb/logical_type.rs]: I think you really want the validation for rust's sake when creating a new one, I can add it to the macro.

Comment (joseph-isaacs) [reply vortex-duckdb/src/duckdb/logical_type.rs]: I see, I didn't understand what you mean before. It doesn't seem that would crazy to have in the macro

File: vortex-duckdb/src/duckdb/logical_type.rs
Comment (robert3005) [reply vortex-duckdb/src/duckdb/logical_type.rs:1]: the only thing here is casting array element size from idx_t (u64) to u32. Arrow also enforces a u32. Not sure what keeping it as idx_t gives us since every call site will have to convert it to u32

File: vortex-duckdb/src/exporter/struct_.rs
Comment (joseph-isaacs) [inline vortex-duckdb/src/exporter/struct_.rs:1]: The last thing to check is if the children need to be flattened? Asked another way can we have children with selection mask or as constant vectors?

Comment (joseph-isaacs) [reply vortex-duckdb/src/exporter/struct_.rs:1]: A few tests checking this would be great.

Comment (robert3005) [reply vortex-duckdb/src/exporter/struct_.rs:1]: Added a test showing constant and dictionary, also added mask implementation for constant. Also added specialisation for maskedvtable to the exporter

File: vortex-duckdb/src/exporter/mod.rs [hunk 11]
```diff
@@ -175,10 +176,7 @@ fn new_array_exporter_with_flatten(
     }
 
     if let Some(array) = array.as_opt::<MaskedVTable>() {
-        return Ok(validity::new_exporter(
-            array.validity_mask(),
-            new_array_exporter_with_flatten(array.child(), cache, flatten)?,
-        ));
+        return masked::new_exporter(array, cache);
     }
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/exporter/mod.rs]: This should be correct?

File: vortex-duckdb/src/exporter/masked.rs [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    array: &MaskedArray,
+    cache: &ConversionCache,
+) -> VortexResult<Box<dyn ColumnExporter>> {
+    // Duckdb doesn't allow constant arrays with validity, we have to flatten the child
+    if let Some(constant_child) = array.child().as_opt::<ConstantVTable>() {
+        constant::new_exporter_with_mask(constant_child, array.validity_mask(), cache)
+    } else {
+        Ok(validity::new_exporter(
+            array.validity_mask(),
+            new_array_exporter(array.child(), cache)?,
```
Comment (joseph-isaacs) [inline vortex-duckdb/src/exporter/masked.rs]: Can we use the `new_array_exporter_with_flatten` method here and ensure that that does in fact flatten the array (if its not all invalid).

Comment (robert3005) [reply vortex-duckdb/src/exporter/masked.rs]: You’re correct this moved the issue one layer further and it still happens

Comment (robert3005) [reply vortex-duckdb/src/exporter/masked.rs]: This whole maskedarray optimisation is pointless, in all the cases where masked array would be helpful to us we can't export it without flattening so we might as well canonicalise

**Comment by robert3005** [comment]
This should be fine now

**Review by 0ax1 — APPROVED** [review]
